### PR TITLE
SQL Database

### PR DESCRIPTION
**Pull request by adzialocha**
* Agnostic SQL database backend via `sqlx` for PostgreSQL, MySQL and SQLite
* Introduce `DATABASE_URL` and `DATABASE_MAX_CONNECTIONS` environment variables
* Manage database via connection pool
* Close connections on shutdown
* Create / use SQLite database in data directory by default
* Embedded database migrations when running application

### CODE REVIEW

File: db/src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        .max_connections(max_connections)
+        .connect(&url)
+        .await?;
+
+    Ok(pool)
+}
+
+/// Run any pending database migrations from inside the application.
+pub async fn run_pending_migrations(pool: &Pool) -> Result<()> {
+    migrate!("./migrations").run(pool).await?;
```
Comment (cafca) [inline db/src/utils.rs:33]: Don't we need the `migrate` cargo feature flag for this?

https://github.com/launchbadge/sqlx#cargo-feature-flags

Comment (cafca) [reply db/src/utils.rs:33]: Could also be interesting for migration management: https://crates.io/crates/sqlx-cli

Comment (cafca) [reply db/src/utils.rs:33]: Also: where is this "./migrations" folder?

**Review by cafca — CHANGES_REQUESTED** [review]
Just a couple of questions - but seriously, please no odd number for the default port!

File: README.md [hunk 2]
```diff
@@ -18,6 +18,8 @@ OPTIONS:
 
 ## Environment variables
 
+* `DATABASE_URL` Database url (SQLite, MySQL, PostgreSQL) (default `sqlite:<data-dir>/main.db`).
+* `DATABASE_MAX_CONNECTIONS` Maximum number of database connections in pool (default `32`).
 * `HTTP_PORT` RPC API HTTP server port (default `9123`).
```
[the remark this answers is not in the document]
Comment (adzialocha) [reply README.md:23]: I haven't thought about the ports yet, we could have something fun maybe? I remember `2020` from a meeting.

One question: What do odd / even ports stand for?

File: db/src/utils.rs (hunk 1 above)
Comment (adzialocha) [reply db/src/utils.rs:33]: > Don't we need the `migrate` cargo feature flag for this?
> 
> https://github.com/launchbadge/sqlx#cargo-feature-flags

It is set by default 🙂

Comment (adzialocha) [reply db/src/utils.rs:33]: > Also: where is this "./migrations" folder?

Its in the `db` folder but not checked in as it's empty, will add a `.gitkeep` 👍

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (cafca) [reply README.md:23]: odd=ugly even=precious

[the remark this answers is not in the document]
Comment (adzialocha) [reply README.md:23]: Also, I found this interesting bit on [Wikipedia](https://en.wikipedia.org/wiki/Network_socket#Definition):

> In the original definition of socket given in RFC 147, as it was related to the ARPA network in 1971, "the socket is specified as a 32 bit number with even sockets identifying receiving sockets and odd sockets identifying sending sockets." Today, however, socket communications are bidirectional.

[the remark this answers is not in the document]
Comment (adzialocha) [reply README.md]: Changed it for now to `2020` (HTTP) and `2022` (WebSocket). Funny to see what is registrated for: https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?&page=36

File: README.md [hunk 3]
```diff
@@ -18,13 +18,13 @@ OPTIONS:
 
 ## Environment variables
 
-* `DATABASE_URL` Database url (SQLite, MySQL, PostgreSQL) (default `sqlite:<data-dir>/main.db`).
+* `DATABASE_URL` Database url (SQLite, MySQL, PostgreSQL) (default `sqlite:<data-dir>/p2panda-node.sqlite3`).
 * `DATABASE_MAX_CONNECTIONS` Maximum number of database connections in pool (default `32`).
-* `HTTP_PORT` RPC API HTTP server port (default `9123`).
+* `HTTP_PORT` RPC API HTTP server port (default `2020`).
```
Comment (cafca) [inline README.md:23]: ❣️

**Review by cafca — APPROVED** [review]

**Pull request merged**